playback/src/audio_backend/alsa.rs

```diff
@@ -2,41 +2,56 @@ use super::{Open, Sink, SinkAsBytes, SinkError, SinkResult};
 use crate::config::AudioFormat;
 use crate::convert::Converter;
 use crate::decoder::AudioPacket;
-use crate::{NUM_CHANNELS, SAMPLE_RATE};
+use crate::{NUM_CHANNELS, SAMPLE_RATE as DECODER_SAMPLE_RATE};
 use alsa::device_name::HintIter;
 use alsa::pcm::{Access, Format, Frames, HwParams, PCM};
 use alsa::{Direction, ValueOr};
 use std::process::exit;
 use thiserror::Error;
 
-const MAX_BUFFER: Frames = (SAMPLE_RATE / 2) as Frames;
-const MIN_BUFFER: Frames = (SAMPLE_RATE / 10) as Frames;
-const ZERO_FRAMES: Frames = 0;
+const OPTIMAL_NUM_PERIODS: Frames = 5;
+const MIN_NUM_PERIODS: Frames = 2;
 
-const MAX_PERIOD_DIVISOR: Frames = 4;
-const MIN_PERIOD_DIVISOR: Frames = 10;
+const COMMON_SAMPLE_RATES: [u32; 14] = [
```

Comment:
Seems these constants should be moved out of Alsa, to be generic over all backends.

Reply:
I can do that although I'm not sure how much use they would get from other backends?

Reply:
Done with the COMMON_SAMPLE_RATES. I left it as an array for now as I don't want to add like 100 more lines of code to make it an enum and impl all the stuff needed to be able to basically use it as an array anyway.

Reply:
That's fine. And if it's too much you can use a macro.

Reply:
Do you want me to create a 14 variant enum and then implement an `as_u32` function (you can't implement a direct `From<Foo> for u32` since u32 is not implemented by us. You'd have to write a wrapper anyway. Or you could make is so it deref'd to a u32) and then implement `IntoIter` only to just use it an as array anyway? THat's what I mean. If it's being used as an array it should just stay an array IMHO.

Reply:
So something like this?:

```rust
use std::convert::TryFrom;
use std::fmt;

enum CommonSampleRates {
    Hz8000,
    Hz11025,
    Hz16000,
    Hz22050,
    Hz44100,
    Hz48000,
    Hz88200,
    Hz96000,
    Hz176400,
    Hz192000,
    Hz352800,
    Hz384000,
    Hz705600,
    Hz768000,
}

impl TryFrom<u32> for CommonSampleRates {
    type Error = ();

    fn try_from(value: u32) -> Result<Self, Self::Error> {
        use CommonSampleRates::*;

        match value {
            8000 => Ok(Hz8000),
            11025 => Ok(Hz11025),
            16000 => Ok(Hz16000),
            22050 => Ok(Hz22050),
            44100 => Ok(Hz44100),
            48000 => Ok(Hz48000),
            88200 => Ok(Hz88200),
            96000 => Ok(Hz96000),
            176400 => Ok(Hz176400),
            192000 => Ok(Hz192000),
            352800 => Ok(Hz352800),
            384000 => Ok(Hz384000),
            705600 => Ok(Hz705600),
            768000 => Ok(Hz768000),
            _ => Err(()),
        }
    }
}

impl fmt::Display for CommonSampleRates {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        use CommonSampleRates::*;

        let rate_str = match self {
            Hz8000 => "8kHz",
            Hz11025 => "11.025kHz",
            Hz16000 => "16kHz",
            Hz22050 => "22.05kHz",
            Hz44100 => "44.1kHz",
            Hz48000 => "48kHz",
            Hz88200 => "88.2kHz",
            Hz96000 => "96kHz",
            Hz176400 => "176.4kHz",
            Hz192000 => "192kHz",
            Hz352800 => "352.8kHz",
            Hz384000 => "384kHz",
            Hz705600 => "705.6kHz",
            Hz768000 => "768kHz",
        };

        write!(f, "{}", rate_str)
    }
}

impl CommonSampleRates {
    fn contains(rate: u32) -> bool {
        const COMMON_SAMPLE_RATES: [u32; 14] = [
            8000, 11025, 16000, 22050, 44100, 48000, 88200, 96000, 176400, 192000, 352800, 384000,
            705600, 768000,
        ];

        COMMON_SAMPLE_RATES.iter().any(|&r| r == rate)
    }
}

```

Reply:
Ok. Done.